packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,227 @@
+    );
+
+    // For each token, extract the price and convert to a wei'd BN. Note that if the type is WETH then the price is 1.
+    // This is done as coingecko does not always return the price of 1 for WETH, in ETH.
+    for (const [index, priceResponse] of tokenPrices.entries()) {
+      this.l1TokenInfo[this.l1Tokens[index]].tokenEthPrice =
+        this.l1TokenInfo[this.l1Tokens[index]].tokenType == TokenType.WETH ? toBNWei("1") : toBNWei(priceResponse[1]);
+    }
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Updated prices",
+      tokenInfo: objectMap(this.l1TokenInfo, (value: { tokenType: TokenType; tokenEthPrice: BN }) => {
+        return { tokenType: TokenType[value.tokenType], tokenEthPrice: fromWei(value.tokenEthPrice) };
+      }),
+    });
+  }
+
+  getRelaySubmitTypeBasedOnProfitability(
+    l1Token: string,
+    cumulativeGasPrice: BN, // sum of the most recent base fee per gas and the current estimated maxPriorityFeePerGas.
+    slowRevenue: BN,
+    speedUpRevenue: BN,
+    instantRevenue: BN
+  ): RelaySubmitType {
+    this._throwIfNotInitialized();
+    if (!this.l1TokenInfo[l1Token]) throw new Error("Token info not found. Ensure to construct correctly");
+    const { tokenType, tokenEthPrice } = this.l1TokenInfo[l1Token];
+
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Considering relay profitability",
+      l1Token,
+      tokenType: TokenType[tokenType],
+      tokenEthPrice: fromWei(tokenEthPrice),
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      slowRevenue: slowRevenue.toString(),
+      speedUpRevenue: speedUpRevenue.toString(),
+      instantRevenue: instantRevenue.toString(),
+    });
+
+    // Calculate the transaction submission costs depending on the gas price and the token type being relayed. This
+    // considers the relayer discount, applied to all relay actions equally.
+    const ethSubmissionCost = this.getRelayEthSubmissionCost(cumulativeGasPrice, tokenType);
+
+    // Calculate the expected revenue, in ETH, based on the amount of tokens being offered in fees.
+    const ethRevenue = this.getEthRevenue(tokenEthPrice, slowRevenue, speedUpRevenue, instantRevenue);
+
+    // Calculate the relay profitability as the difference between revenue and cost.
+    const ethProfitability = this.getProfit(ethSubmissionCost, ethRevenue);
+
+    // Finally, determine the relay submit type based on the profitability.
+    let relaySubmitType: RelaySubmitType;
+    if (
+      ethProfitability.instantEthProfit.gt(toBN(0)) &&
+      ethProfitability.instantEthProfit.gte(ethProfitability.speedUpEthProfit) &&
+      ethProfitability.instantEthProfit.gte(ethProfitability.slowEthProfit)
+    )
+      relaySubmitType = RelaySubmitType.Instant;
+    else if (
+      ethProfitability.speedUpEthProfit.gt(toBN(0)) &&
+      ethProfitability.speedUpEthProfit.gt(ethProfitability.slowEthProfit)
```

Comment:
Should this be a `gte` too?

Reply:
Also, how can there be a positive `speedUpEthProfit` AND a positive `slowEthProfit`? I.e., a speed up relay should only be possible if a slow relay is pending, and therefore a slow relay should be impossible?

Reply:
ye, it should be gte. it's really pretty immaterial here though as these are wei numbers.

regarding how there could be a positive speedUpEthProfit AND a positive slowEthProfit, I agree that logically that should not happen due to how the revenues are calculated.

would you agree that this whole block can be reduced to this:

```ts
if (
      ethProfitability.instantEthProfit.gt(toBN(0)) &&
      ethProfitability.instantEthProfit.gte(ethProfitability.speedUpEthProfit) &&
      ethProfitability.instantEthProfit.gte(ethProfitability.slowEthProfit)
    )
      relaySubmitType = RelaySubmitType.Instant;
    else if (ethProfitability.speedUpEthProfit.gt(toBN(0))) relaySubmitType = RelaySubmitType.SpeedUp;
    else if (ethProfitability.slowEthProfit.gt(toBN(0))) relaySubmitType = RelaySubmitType.Slow;
    else relaySubmitType = RelaySubmitType.Ignore;
```

Reply:
also note that the `shouldRelay` method would have set the `speedUpRevenue` to 0 in the event that the relay already had been slow relayed.

Reply:
in fact after thinking about this more I think we should only instant relay iff the profit is greater than (not greater than or equal) to the slow eth profit. Reason is if these are equal then the depositor has set the instant relay fee to 0 and in this case the relayer should do a slow relay.

I also modified the `shouldRelay` method. it now has:

`instantRevenue = hasInstantRelayer ? toBN(0) : slowRevenue.add(speedUpRevenue);` such that we only generate an instant revenue if the relay has not been instant relayed.

Reply:
Ok I agree with this